_posts/2019-01-17-multi-store-custom-telemetry-with-shared-data.md

```diff
@@ -0,0 +1,207 @@
+When fired, `notify()` will be called, which then collects the payload from `custom-store` and schedule the ping for sending.
+Telemetry Core will take care of actually sending the ping (and retrying, storing it in the archive, etc.).
+For test reasons only we also send this ping when `start()` is called (so we don't actually have to wait 24 hours to see something).
+
+```javascript
+var EXPORTED_SYMBOLS = ["TelemetryCustomPing"];
+
+var TelemetryCustomPing = Object.freeze({
+  start() {
+    UpdateTimerManager.registerTimer(
+      "telemetry_custom_ping",
+      this,
+      24 * 60 * 60; /* 1 day */
+    );
+
+    this.sendPing();
+  },
+
+  notify() {
+    this.sendPing();
+  },
+
+  sendPing() {
+    // Let's record just one more value, so _something_ is included.
+    let randValue = Math.floor(Math.random() * Math.floor(100));
+    Services.telemetry.getHistogramById("TELEMETRY_TICK_TIMES_RAND").add(randValue);
+
+    // We only include the histograms, as that's all we are recording for this ping.
+    let payload = {
+      custom: "This is custom data",
+      histograms: Services.telemetry.getSnapshotForHistograms("custom-store", /* clear */ true),
+    };
+
+    TelemetryController.submitExternalPing("custom", payload,
+      {
+        addClientId: false,
+        addEnvironment: true,
+      }
+    );
+  }
+});
+```
+
+Now we need to actually start this ping timer. This should be done in the initialization phase of `TelemetryController`:
+
+```patch
+--- c/toolkit/components/telemetry/app/TelemetryController.jsm
++++ i/toolkit/components/telemetry/app/TelemetryController.jsm
+@@ -62,6 +62,7 @@ XPCOMUtils.defineLazyModuleGetters(this, {
+   TelemetryReportingPolicy: "resource://gre/modules/TelemetryReportingPolicy.jsm",
+   TelemetryModules: "resource://gre/modules/ModulesPing.jsm",
+   TelemetryUntrustedModulesPing: "resource://gre/modules/UntrustedModulesPing.jsm",
++  TelemetryCustomPing: "resource://gre/modules/CustomPing.jsm",
+   UpdatePing: "resource://gre/modules/UpdatePing.jsm",
+   TelemetryHealthPing: "resource://gre/modules/HealthPing.jsm",
+   TelemetryEventPing: "resource://gre/modules/EventPing.jsm",
+@@ -720,6 +721,9 @@ var Impl = {
+           if (AppConstants.NIGHTLY_BUILD && AppConstants.platform == "win") {
+             TelemetryUntrustedModulesPing.start();
+           }
++
++          // Start the custom ping, which reports minimal information.
++          TelemetryCustomPing.start();
+         }
+
+         TelemetryEventPing.startup();
+```
+
+**But it's not measuring anything!**
+
+That is right. Let's add some random values once in a while:
+
+```patch
+--- c/toolkit/components/telemetry/pings/TelemetrySession.jsm
++++ i/toolkit/components/telemetry/pings/TelemetrySession.jsm
+@@ -440,6 +440,10 @@ var TelemetryScheduler = {
+   _schedulerTickLogic() {
+     this._log.trace("_schedulerTickLogic");
+
++    // Adding a random value at every tick.
```

Comment:
I'd avoid to publish an example with a change to `TelemetrySession.jsm`. Why would you need this anyway? You're populating the histogram in the custom ping file.